### PR TITLE
Add pull-through caching registry mirrors for CI tests

### PR DESCRIPTION
**Pull request by brandond**
#### Proposed Changes ####

Add pull-through caching registry mirrors for CI tests

#### Types of Changes ####

* CI

#### Verification ####

* Create PR, merge PR, create tag, etc. Note lack of image pull rate-limit

#### Linked Issues ####

k3s-io/k3s#2684 
k3s-io/k3s#2656

#### Further Comments ####

Adds pull-though cache for: docker.io k8s.gcr.io gcr.io quay.io ghcr.io
Each cache listens on a separate port and has an explicit entry in registries.yaml

### CODE REVIEW

File: scripts/test-helpers [hunk 1]
```diff
@@ -478,6 +480,31 @@ export -f provision-cluster
 
 # ---
 
+provision-registry-proxy() {
+    set -e -o pipefail
+    local image="docker.io/rpardini/docker-registry-proxy:0.6.1"
```
Comment (Oats87) [inline scripts/test-helpers:485]: Should we add this to our image mirror?

Comment (brandond) [reply scripts/test-helpers:487]: Yeah probably, once I get it working...

File: scripts/test-helpers [hunk 2]
```diff
@@ -480,6 +486,50 @@ export -f provision-cluster
...
+        local status=$(docker inspect $name --format '{{ .State.Status }} {{ .Config.Image }} {{ (index .HostConfig.PortBindings "5000/tcp" 0).HostPort }}' 2>/dev/null || true)
+
+        read state_status config_image hostport <<< $status
+        if [ "$state_status" != "running" ] || [ "$config_image" != "$image" ]; then
+            hostport=$(timeout --foreground 5s bash -c get-port)
+            docker rm --force $name 2>/dev/null || true
+            docker run \
+                -d --name $name \
+                -p 0.0.0.0:$hostport:5000 \
+                -v "registry-cache:/var/lib/registry" \
```
Comment (dweomer) [inline scripts/test-helpers:514]: A shared cache will invite stomping for lower layers that are common between registries (I am thinking of re-tagged/pushed items of we we use a few).

Comment (dweomer) [inline scripts/test-helpers:514]:
```suggestion
                -v "registry-data-$name:/var/lib/registry" \
```

Comment (brandond) [reply scripts/test-helpers:514]: The caches are only used for the test clusters, which don't push/pull any of the k3s images. All the building and tagging is done at this point.

Comment (brandond) [reply scripts/test-helpers:514]: I'd prefer to use just one volume. The individual registries are in separate subdirectories of this volume (see the env var)

Comment (brandond) [reply scripts/test-helpers:514]: Also, they actual cache is in a separate directory: `"REGISTRY_STORAGE_FILESYSTEM_ROOTDIRECTORY=/var/lib/registry/$registry_name"`

Comment (dweomer) [reply scripts/test-helpers:514]: And multiple test clusters may be running at once and pulling images with duplicated layers from different locations, thus, stompage.

Comment (dweomer) [reply scripts/test-helpers:514]: ah, i missed ` -e "REGISTRY_STORAGE_FILESYSTEM_ROOTDIRECTORY=/var/lib/registry/$registry_name" \`

**Review by dweomer — APPROVED** [review]

**Review by Oats87 — APPROVED** [review]

File: scripts/test-helpers [hunk 3]
```diff
@@ -480,6 +486,50 @@ export -f provision-cluster
 
 # ---
 
+provision-registry-proxy() {
+    set -e -o pipefail
```
Comment (erikwilson) [inline scripts/test-helpers:490]: pipefail is probably a good thing to have in general, not sure if anything depends on that. if it is called that way in this function it has a global effect tho. is there anything particular in this function that requires it?

Comment (brandond) [reply scripts/test-helpers:490]: I think I've refactored out anything that might have needed it in an earlier version. I can remove it if you think it'll cause problems?

**Pull request merged**